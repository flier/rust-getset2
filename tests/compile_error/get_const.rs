use getset2::Getter;

#[derive(Getter)]
pub struct Struct {
    #[get(const, clone)] // #[get(clone)] is not allowed to a `const` getter
    field: Foobar,
}

#[derive(Getter)]
#[get(const, clone)] // #[get(clone)] is not allowed to a `const` getter
pub struct Struct2 {
    field: Foobar,
}

#[derive(Clone)]
pub struct Foobar;

fn main() {}
