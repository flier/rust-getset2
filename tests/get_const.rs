use getset2::Getter;

#[derive(Default, Getter)]
#[get(pub, const, mut)]
struct Foo {
    /// `pub const fn private_field(&self) -> &usize`
    private_field: usize,

    /// `pub const fn copy_field(&self) -> usize`
    #[get(copy)]
    copy_field: usize,

    /// `pub fn clone_field(&self) -> usize`
    #[get(clone, const(false))] // #[get(clone)] is not allowed to a `const` getter
    clone_field: usize,
}

#[test]
fn get_const() {
    let mut foo = Foo::default();

    *foo.private_field_mut() = 123;
    assert_eq!(foo.private_field(), &123);

    *foo.copy_field_mut() = 456;
    assert_eq!(foo.copy_field(), 456);

    *foo.clone_field_mut() = 789;
    assert_eq!(foo.clone_field(), 789);
}
