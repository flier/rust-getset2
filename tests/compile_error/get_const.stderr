error[E0015]: cannot call non-const fn `<Foobar as Clone>::clone` in constant functions
 --> tests/compile_error/get_const.rs:5:5
  |
5 |     #[get(const, clone)] // #[get(clone)] is not allowed to a `const` getter
  |     ^
  |
  = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants

error[E0015]: cannot call non-const fn `<Foobar as Clone>::clone` in constant functions
  --> tests/compile_error/get_const.rs:10:1
   |
10 | #[get(const, clone)] // #[get(clone)] is not allowed to a `const` getter
   | ^
   |
   = note: calls in constant functions are limited to constant functions, tuple structs and tuple variants
